// react-social-auth/lib/google.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.onClick = exports.init = void 0;

var _fp = _interopRequireDefault(require("lodash/fp"));

var _common = require("./common");

var _bluebird = _interopRequireDefault(require("bluebird"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var getAuthPayload = function getAuthPayload(appId, user) {
  var profile = user.getBasicProfile();
  var authResponse = user.getAuthResponse();
  return {
    type: 'google',
    profile: {
      id: profile.getId(),
      firstName: profile.getGivenName(),
      lastName: profile.getFamilyName(),
      email: profile.getEmail()
    },
    authResponse: {
      clientId: appId,
      token: authResponse.id_token
    }
  };
};

var init = function init(_ref) {
  var appId = _ref.appId,
      _ref$scope = _ref.scope,
      scope = _ref$scope === void 0 ? 'profile email' : _ref$scope,
      _ref$fetchBasicProfil = _ref.fetchBasicProfile,
      fetchBasicProfile = _ref$fetchBasicProfil === void 0 ? true : _ref$fetchBasicProfil;
  (0, _common.loadScript)('google-platform', 'https://apis.google.com/js/platform.js').then(function () {
    var load = _bluebird["default"].promisify(window.gapi.load);

    load('auth2').then(function () {
      if (!window.gapi.auth2.getAuthInstance()) {
        window.gapi.auth2.init({
          client_id: _fp["default"].trimEnd(appId, '.apps.googleusercontent.com'),
          fetch_basic_profile: fetchBasicProfile,
          scope: scope
        });
      }
    });
  });
};

exports.init = init;

var onClick = function onClick(_ref2) {
  var appId = _ref2.appId,
      onSuccess = _ref2.onSuccess;
  var auth2 = window.gapi.auth2.getAuthInstance();

  if (auth2.isSignedIn && auth2.isSignedIn.get()) {
    var user = auth2.currentUser.get();
    onSuccess(getAuthPayload(appId, user));
  } else {
    auth2.signIn().then(function (user) {
      onSuccess(getAuthPayload(appId, user));
    });
  }
};

exports.onClick = onClick;
