// react-social-auth/lib/linkedin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.onClick = exports.init = void 0;

var _common = require("./common");

var _browser = _interopRequireDefault(require("simple-random/browser"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var init = function init(_ref) {
  var appId = _ref.appId,
      onSuccess = _ref.onSuccess;

  if (localStorage.linkedInLoginState) {
    var redirectUri = localStorage.linkedInReactLoginRedirectUri;
    var code = (0, _common.getQueryParameter)('code');
    var state = (0, _common.getQueryParameter)('state');
    window.history.replaceState(null, null, redirectUri);

    if (code && state === localStorage.linkedInLoginState) {
      delete localStorage.linkedInLoginState;
      onSuccess({
        type: 'linkedin',
        authResponse: {
          grant_type: 'authorization_code',
          code: code,
          client_id: appId,
          redirect_uri: redirectUri
        }
      });
    }
  }
};

exports.init = init;

var onClick = function onClick(_ref2) {
  var appId = _ref2.appId,
      _ref2$scope = _ref2.scope,
      scope = _ref2$scope === void 0 ? 'r_basicprofile r_emailaddress' : _ref2$scope,
      _ref2$redirectUri = _ref2.redirectUri,
      redirectUri = _ref2$redirectUri === void 0 ? window.location.href : _ref2$redirectUri;
  localStorage.linkedInReactLoginRedirectUri = redirectUri;
  localStorage.linkedInLoginState = (0, _browser["default"])({
    length: 8
  });
  window.location.href = "https://www.linkedin.com/oauth/v2/authorization?response_type=code&client_id=".concat(appId, "&redirect_uri=").concat(encodeURIComponent(redirectUri), "&state=").concat(localStorage.linkedInLoginState, "&scope=").concat(encodeURIComponent(scope));
};

exports.onClick = onClick;
