// react-social-auth/lib/salesforce.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.onClick = exports.init = void 0;

var _common = require("./common");

var _browser = _interopRequireDefault(require("simple-random/browser"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var init = function init(_ref) {
  var onSuccess = _ref.onSuccess;

  if (localStorage.salesforceLoginState) {
    var redirectUri = localStorage.salesforceReactLoginRedirectUri;
    var code = (0, _common.getQueryParameter)('code');
    var state = (0, _common.getQueryParameter)('state');
    window.history.replaceState(null, null, redirectUri);

    if (code && state === localStorage.salesforceLoginState) {
      delete localStorage.salesforceLoginState;
      onSuccess({
        type: 'salesforce',
        authResponse: {
          code: code,
          redirectUri: redirectUri
        }
      });
    }
  }
};

exports.init = init;

var onClick = function onClick(_ref2) {
  var appId = _ref2.appId,
      _ref2$redirectUri = _ref2.redirectUri,
      redirectUri = _ref2$redirectUri === void 0 ? window.location.href : _ref2$redirectUri;
  localStorage.salesforceReactLoginRedirectUri = redirectUri;
  localStorage.salesforceLoginState = (0, _browser["default"])({
    length: 8
  });
  window.location.href = "https://login.salesforce.com/services/oauth2/authorize?response_type=code&client_id=".concat(appId, "&redirect_uri=").concat(redirectUri, "&state=").concat(localStorage.salesforceLoginState);
};

exports.onClick = onClick;
